packages/composite-checkout/src/components/billing-fields.js

```diff
@@ -0,0 +1,552 @@
+	margin-top: 16px;
+	:first-child {
+		margin-top: 0;
+	}
+`;
+
+const FieldRow = styled( GridRow )`
+	margin-top: 16px;
+
+	first-child {
+		margin-top: 0;
+	}
+`;
+
+function hasDomainsInCart( items ) {
+	if ( items.find( item => item.type === 'domain' ) ) {
+		return true;
+	}
+
+	return false;
+}
+
+function isElligibleForVat() {
+	//Detect whether people are in EU or AU and return true if they are
+	return false;
+}
+
+function DomainFieldsCheckbox( toggleVisibility ) {
+	return (
+		<DomainRegistrationCheckBoxWrapper>
+			<DomainRegistrationCheckbox
+				type="checkbox"
+				id="domain-registration"
+				name="domain-registration"
+				defaultChecked={ true }
+				onChange={ toggleVisibility }
+			/>
+			<DomainRegistrationLabel htmlFor="domain-registration">
+				Use your billing details for your domain registration contact information.
+			</DomainRegistrationLabel>
+		</DomainRegistrationCheckBoxWrapper>
+	);
+}
+
+DomainFieldsCheckbox.propTypes = {
+	toggleVisibility: PropTypes.func,
+};
+
+const DomainRegistrationCheckBoxWrapper = styled.div`
+	margin: 16px 0 24px;
+	display: flex;
+	width: 100%;
+`;
+
+const DomainRegistrationLabel = styled.label`
+	font-size: 14px;
+	color: ${props => props.theme.colors.textColor};
+	display: block;
+	position: relative;
+	padding-left: 5px;
+
+	:hover {
+		cursor: pointer;
+	}
+
+	:before {
+		display: block;
+		width: 16px;
+		height: 16px;
+		content: '';
+		position: absolute;
+		left: -20px;
+		top: 0;
+		border: 1px solid ${props => props.theme.colors.borderColorLight};
+		border-radius: 3px;
+	}
+`;
+
+const DomainRegistrationCheckbox = styled.input`
+	margin-right: 5px;
+	opacity: 0;
+
+	:checked + label:before {
+		background: ${props => props.theme.colors.highlight};
+	}
+
+	:checked + label:after {
+		display: block;
+		width: 4px;
+		height: 8px;
+		content: '';
+		position: absolute;
+		left: -14px;
+		top: 3px;
+		border-right: 2px solid ${props => props.theme.colors.surface};
+		border-bottom: 2px solid ${props => props.theme.colors.surface};
+		transform: rotate( 45deg );
+	}
+
+	:focus + label:before {
+		outline: ${props => props.theme.colors.outline} auto 5px;
+	}
+`;
+
+function AddressFields( { paymentData, setPaymentData, location } ) {
+	const localize = useLocalize();
+	const name = location + 'Name';
+	const address = location + 'Address';
+	const city = location + 'City';
+	const state = location + 'State';
+	const province = location + 'Province';
+
+	const {
+		billingName = '',
+		billingAddress = '',
+		billingCity = '',
+		billingState = '',
+		billingProvince = '',
+		billingNameError = null,
+		billingAddressError = null,
+		billingCityError = null,
+		billingStateError = null,
+		billingProvinceError = null,
+		domainsName = '',
+		domainsAddress = '',
+		domainsCity = '',
+		domainsState = '',
+		domainsProvince = '',
+		domainsNameError = null,
+		domainsAddressError = null,
+		domainsCityError = null,
+		domainsStateError = null,
+		domainsProvinceError = null,
```

Comment:
Is there any way we can dynamically render these values using the location prop before the JS compiler gets to it?

I was able to dynamically write some of the other parts of this component, it would be nice if I could clean this area up too.

Reply:
Creating object keys by joining strings together is a risky pattern. 

Let's say that in the future we encounter a bug and the error message shows that the property `domainsName` is not set. To debug this issue we'll search the code for the string `domainsName`, but we won't find it (well, in this case we will because it's also a variable you created above but that's not the same as the object key and so we might trace the wrong thing).

It's much safer to be explicit ("[explicit is better than implicit](https://en.wikipedia.org/wiki/Zen_of_Python)" is a good guideline) rather than trying to create things dynamically. The old checkout form creates a lot of its fields using a dynamic `createField` function, but that has created all kinds of issues.

That said, we can create safe, explicit abstractions to simplify creating variables and fields. For example, something like this would get rid of this big destructure and simplify the change functions:

```js
const location = 'domains';
//...
const currentPaymentData = paymentData || {};
const currentLocationData = currentPaymentData[ location ] || {};
const updatePaymentData = ( key, value ) => setPaymentData( { ...currentPaymentData, [ key ]: value } );
const updateLocationData = ( key, value ) => updatePaymentData( { ...currentLocationData, [ key ]: value } );
const updateName = value => updateLocationData( 'name ', value );
const updateAddress = value => updateLocationData( 'address', value );
//...
<FormField value={ currentLocationData.name || '' } error={ currentLocationData.nameError } onChange={ updateName } />
```

That said, I wonder if it might be a good idea to change the way we store payment data so that it uses a reducer function instead. That way we could handle the object merging there and change the above to this:

```js
const location = 'domains';
//...
const currentPaymentData = paymentData || {};
const currentLocationData = currentPaymentData[ location ] || {};
const updateName = value => updateLocationPaymentData( { location, key: 'name', value } );
const updateAddress = value => updateLocationPaymentData( { location, key: 'address', value } );
//...
<FormField value={ currentLocationData.name || '' } error={ currentLocationData.nameError } onChange={ updateName } />
```

In either form, we could inline the `onChange` function to avoid some lines of code:

```js
<FormField value={ currentLocationData.name || '' } error={ currentLocationData.nameError } onChange={ value => updateLocationData( 'name ', value ) } />
```

Reply:
Also, we might not want to store the validation errors in the payment data state; are they needed in other steps? I don't think they are. In that case we could create a separate local state for this component that just holds the errors.

Reply:
Thanks this helps. Both options you shared look great. I'll try them out and see how it goes.

> Also, we might not want to store the validation errors in the payment data state; are they needed in other steps? I don't think they are. In that case we could create a separate local state for this component that just holds the errors.

This is a good point. We're currently not doing anything with the validation so maybe I should take them out for now and we can add them back as a separate PR.

Reply:
Alright, I got this working now but I'm not 100% sure what it's doing and I duplicated some code across a number of methods. Would it make sense to move the following code to a helper method:

```
	const currentPaymentData = paymentData || {};
	const currentLocationData = currentPaymentData[ fieldType ] || {};
	const updatePaymentData = ( key, value ) => setPaymentData( { ...currentPaymentData, [ key ]: value } );
	const updateLocationData = ( key, value ) => updatePaymentData( fieldType, { ...currentLocationData, [ key ]: value } );
```